### PR TITLE
Add support for runtime builder versioning 

### PR DESCRIPTION
**Pull request by nkubala**
This PR enables runtime maintainers to now publish a set of versioned runtime builders. Maintainers can now specify a list of "aliases", which will resolve to one of the versioned builders. This will all be specified through a `runtime.yaml` manifest file, which will be checked in to version control.

Example manifest:

```
runtimes:
  java-jetty-9:
    target:
      file: gs://runtime-builders/java-jetty-9-1234.yaml
  java-openjdk-8:
    target:
      file: gs://runtime-builders/java-openjdk-8-1234.yaml
  java:
    target:
      runtime: java-openjdk
  java-openjdk:
    target:
      runtime: java-openjdk-8
  java-jetty:
    target:
      runtime: java-jetty-9

  python:
    target:
      runtime: python2
  python2:
    target:
      runtime: python2.7
  python2.7:
    target:
      file: gs://runtime-builders/python27-1234.yaml
  python3:
    target:
      runtime: python3.4
  python3.4:
    target:
      file: gs://runtime-builders/python3-1234.yaml
```

This manifest file will be verified before publishing to make sure there are no dead/circular aliases provided by the maintainers.

@znewman01 @markpell @dlorenc

### CODE REVIEW

File: runtime_builders/builder_util.py [hunk 1]
```diff
@@ -36,11 +36,13 @@ def write_to_gcs(gcs_path, file_contents):
...
 
         command = ['gsutil', 'cp', f_name, gcs_path]
+        output = ''
         try:
             output = subprocess.check_output(command)
         except subprocess.CalledProcessError as e:
-            logging.error('Error encountered when writing to GCS!: {0}'
-                          .format(output))
+            logging.error('Error encountered when writing to GCS!')
+            if output is not '':
```
Comment (dlorenc) [inline runtime_builders/builder_util.py]: I think you can just say "if output" here.

Comment (dlorenc) [inline runtime_builders/builder_util.py]: Actually I don't think there's any chance output will be filled in here if an exception gets thrown, so you can probably just drop this.

To get the output in the error case, you have to get it off the exception object:
https://docs.python.org/2/library/subprocess.html#subprocess.check_output

File: runtime_builders/builder_util.py [hunk 2]
```diff
@@ -57,10 +59,51 @@ def get_file_from_gcs(gcs_file, temp_file):
...
+    Aliases are provided for runtime 'names' that can be included in users'
+    application configuration files: this method ensures that all the aliases
+    can resolve to actual builder files.
+
+    All builders and aliases are turned into nodes in a graph, which is then
+    traversed to be sure that all nodes lead down to a builder node.
+    """
+    try:
+        node_graph = {}
+        for item in manifest.get('runtimes').items():
```
Comment (dlorenc) [inline runtime_builders/builder_util.py]: This might be a little cleaner with `for key, val in manifest.get('runtimes').iteritems()'

File: runtime_builders/builder_util.py [hunk 3]
```diff
@@ -57,10 +59,51 @@ def get_file_from_gcs(gcs_file, temp_file):
...
 
-def write_manifest_file(manifest):
+def verify_and_write_manifest(manifest):
+    """Verify that the provided runtime manifest is valid before publishing.
+
+    Aliases are provided for runtime 'names' that can be included in users'
+    application configuration files: this method ensures that all the aliases
+    can resolve to actual builder files.
+
+    All builders and aliases are turned into nodes in a graph, which is then
```
Comment (dlorenc) [inline runtime_builders/builder_util.py]: Could you include the expected format here in the comment?

File: runtime_builders/builder_util.py [hunk 4]
```diff
@@ -57,10 +59,51 @@ def get_file_from_gcs(gcs_file, temp_file):
         return False
 
 
-def write_manifest_file(manifest):
+def verify_and_write_manifest(manifest):
```
Comment (dlorenc) [inline runtime_builders/builder_util.py]: Could you split this into a verify method and a write method? Then you could add a set of unit tests for the verify portion.[

File: runtime_builders/builder_util.py (hunk 1 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: hmm yeah not sure what I was thinking here, the output that I care about should actually go out to stdout here I think (based on a test I just did). going to remove this regardless

File: runtime_builders/builder_util.py (hunk 4 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: yeah good call, then we can run the verification as a presubmit before the manifest gets copied.

File: runtime_builders/builder_util.py (hunk 3 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: done

File: runtime_builders/builder_util.py (hunk 2 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: :+1:

File: runtime_builders/builder_util.py [hunk 5]
```diff
@@ -57,10 +63,57 @@ def get_file_from_gcs(gcs_file, temp_file):
...
+        target:
+          runtime: java-openjdk
+      java-openjdk:
+        target:
+          file: gs://runtimes/java-openjdk-1234.yaml
+    """
+    try:
+        node_graph = {}
+        for key, val in manifest.get('runtimes').iteritems():
+            if key == 'schema_version':
```
Comment (markpell) [inline runtime_builders/builder_util.py]: schema version is a top level entry, not a key in runtimes

File: runtime_builders/builder_util.py [hunk 6]
```diff
@@ -57,10 +63,57 @@ def get_file_from_gcs(gcs_file, temp_file):
...
+                continue
+            target = val['target']
```
Comment (markpell) [inline runtime_builders/builder_util.py]: This will crash if there is no target.  Technically this is valid if we remove a runtime.  We would potentially want to leave it in the manifest because we could keep a deprecation message there.  In the client it would print the message, but fail because the runtime no longer exists.

File: runtime_builders/builder_util.py [hunk 7]
```diff
@@ -57,10 +63,57 @@ def get_file_from_gcs(gcs_file, temp_file):
         return False
 
 
-def write_manifest_file(manifest):
-    manifest_contents = yaml.round_trip_dump(manifest,
-                                             default_flow_style=False)
-    write_to_gcs(MANIFEST_FILE, SCHEMA_LINE + manifest_contents)
+def verify_manifest(manifest):
```
Comment (markpell) [inline runtime_builders/builder_util.py]: If you need additional inspiration you can look at runtime_builders.py:361 where I do similar parsing

File: runtime_builders/builder_util.py (hunk 5 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: Ah whoops, meant to remove that.

File: runtime_builders/builder_util.py (hunk 6 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: Ah yeah I forgot about the deprecations. Fixed

File: runtime_builders/builder_util.py [hunk 8]
```diff
@@ -57,10 +64,64 @@ def get_file_from_gcs(gcs_file, temp_file):
...
+            child = None
+            isBuilder = 'file' in target.keys()
+            if not isBuilder:
+                child = target['runtime']
+            node = node_graph.get(key, {})
+            if not node:
+                node_graph[key] = Node(key, isBuilder, child)
+        for _, node in node_graph.items():
+            child = node
+            while True:
```
Comment (dlorenc) [inline runtime_builders/builder_util.py]: Should we add a limit to how deep we'll go here?

File: runtime_builders/template_builder.py [hunk 9]
```diff
@@ -48,10 +48,17 @@ def _resolve_and_publish(config_file, bucket):
...
         with open(config_file, 'r') as f:
-            project_cfg = json.load(f)
+            if 'json' in config_file:
+                project_cfg = json.load(f)
+            elif 'yaml' in config_file:
+                project_cfg = yaml.round_trip_load(f)
+            else:
+                logging.error('Please provide a valid yaml/json config file.')
+                sys.exit(1)
+            print yaml.round_trip_dump(project_cfg, indent=2)
```
Comment (dlorenc) [inline runtime_builders/template_builder.py]: logging?

File: runtime_builders/publish_builders.py [hunk 10]
```diff
@@ -59,38 +58,21 @@ def main():
...
-            logging.warn('Overwriting old {0} builder: {1}'
-                         .format(project_name, prev_builder))
-        m_project['target']['file'] = full_latest_file
-        _write_version_file(project_name, latest_file)
-    except KeyError as ke:
-        logging.error('FATAL: Formatting issue encountered in manifest. '
-                      'Exiting. \n{0}'.format(ke))
-        sys.exit(1)
+def _publish_latest(builder_dir):
+    for file in glob.glob(os.path.join(builder_dir, '*.json')):
```
Comment (dlorenc) [inline runtime_builders/publish_builders.py]: file is reserved, could you use f?

Comment (dlorenc) [inline runtime_builders/publish_builders.py]: Should this be yaml?

File: runtime_builders/builder_util.py (hunk 8 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: Hmm, well one thing I should definitely do is at least keep track of what I've seen, so if there's a circular dependency this doesn't infinite loop

File: runtime_builders/template_builder.py (hunk 9 above)
Comment (nkubala) [inline runtime_builders/template_builder.py]: whoops, I need to stop leaving these in here haha

File: runtime_builders/publish_builders.py (hunk 10 above)
Comment (nkubala) [inline runtime_builders/publish_builders.py]: :+1:

Comment (nkubala) [inline runtime_builders/publish_builders.py]: Since these are already checked in as <runtime>.json, I figured I would just leave them the way they are, unless you strongly prefer that I change them

File: runtime_builders/builder_util.py [hunk 11]
```diff
@@ -57,10 +65,70 @@ def get_file_from_gcs(gcs_file, temp_file):
...
+                continue
+            child = None
+            isBuilder = 'file' in target.keys()
+            if not isBuilder:
+                child = target['runtime']
+            node = node_graph.get(key, {})
+            if not node:
+                node_graph[key] = Node(key, isBuilder, child)
+        for _, node in node_graph.items():
+            seen = Set()
```
Comment (dlorenc) [inline runtime_builders/builder_util.py]: Is this different than the set() builtin?

File: runtime_builders/builder_util.py [hunk 12]
```diff
@@ -57,10 +65,70 @@ def get_file_from_gcs(gcs_file, temp_file):
...
+          file: gs://runtimes/java-openjdk-1234.yaml
+        deprecation:
+          message: "openjdk is deprecated."
+    """
+    try:
+        node_graph = {}
+        for key, val in manifest.get('runtimes').iteritems():
+            target = val.get('target', {})
+            if not target:
+                deprecation = val.get('deprecation', {})
```
Comment (dlorenc) [inline runtime_builders/builder_util.py]: You don't actually use deprecation here, so you could shorten this to:

if not 'deprecation' in val:

File: runtime_builders/builder_util.py [hunk 13]
```diff
@@ -57,10 +65,70 @@ def get_file_from_gcs(gcs_file, temp_file):
...
+        target:
+          runtime: java-openjdk
+      java-openjdk:
+        target:
+          file: gs://runtimes/java-openjdk-1234.yaml
+        deprecation:
+          message: "openjdk is deprecated."
+    """
+    try:
+        node_graph = {}
```
Comment (dlorenc) [inline runtime_builders/builder_util.py]: It looks like this is mainly two parts, building a node graph from yaml and then validating it. Could you separate these into two functions? It might make it easier to read.

File: runtime_builders/template_builder.py [hunk 14]
```diff
@@ -48,10 +48,16 @@ def _resolve_and_publish(config_file, bucket):
     try:
         gcs_paths = []
         with open(config_file, 'r') as f:
-            project_cfg = json.load(f)
+            if 'json' in config_file:
```
Comment (dlorenc) [inline runtime_builders/template_builder.py]: Do we really need to support both yaml and json?

File: runtime_builders/builder_util.py (hunk 13 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: sure, done

File: runtime_builders/builder_util.py (hunk 12 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: good call

File: runtime_builders/builder_util.py (hunk 11 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: ah wow, didn't realize that `sets.Set()` was deprecated. updated

File: runtime_builders/template_builder.py (hunk 14 above)
Comment (nkubala) [inline runtime_builders/template_builder.py]: Not really, I just figured it didn't require much work to support both, to give people the option in case they really felt strongly about one or the other.

Comment (dlorenc) [inline runtime_builders/template_builder.py]: I'd rather stick with just one for simplicity unless we've actually gotten requests for both.

File: runtime_builders/publish_builders.py [hunk 15]
```diff
@@ -15,9 +15,11 @@
 # limitations under the License.
 
 import argparse
+import glob
 import json
 import logging
 import os
+from ruamel import yaml
```
Comment (dlorenc) [inline runtime_builders/publish_builders.py]: Use standard yaml?

File: runtime_builders/builder_util.py [hunk 16]
```diff
@@ -69,8 +144,19 @@ def load_manifest_file():
         command = ['gsutil', 'cp', MANIFEST_FILE, tmp]
```
Comment (dlorenc) [inline runtime_builders/builder_util.py]: "gsutil cat" might be a bit easier to use here, you could avoid writing a tmp file.

File: runtime_builders/builder_util.py [hunk 17]
```diff
@@ -26,7 +27,19 @@
...
 SCHEMA_VERSION = 1
-SCHEMA_LINE = 'schema_version: {0}\n'.format(SCHEMA_VERSION)
+
+
+def copy_to_gcs(file_path, gcs_path):
+    command = ['gsutil', 'cp', file_path, gcs_path]
+    try:
+        output = subprocess.check_output(command)
+        logging.debug(output)
+    except subprocess.CalledProcessError:
```
Comment (dlorenc) [inline runtime_builders/builder_util.py]: Can you log the exception?

File: runtime_builders/publish_builders.py [hunk 18]
```diff
@@ -27,30 +29,24 @@ def main():
...
-                latest = config['latest']
-                if not latest.startswith(prefix):
-                    logging.error('Please provide fully qualified path to '
-                                  'config file in GCS!')
-                    logging.error('Path should start with \'{0}\''
-                                  ''.format(prefix))
-                    sys.exit(1)
-                _format_and_write(config['project'], latest, prefix, manifest)
-        builder_util.write_manifest_file(manifest)
+        if not args.manifest.endswith('.yaml'):
```
Comment (dlorenc) [inline runtime_builders/publish_builders.py]: I think the "right" way to do this with argparse is to pass a "type=" parameter to the add_argument call that does validation and returns the parsed flag, but it's not a huge deal. You could even just drop this check since you'd blow up below if you can't parse/validate the yaml.

File: runtime_builders/publish_builders.py [hunk 19]
```diff
@@ -59,38 +55,21 @@ def main():
...
 
+        latest = config['latest']
+        project_name = config['project']
 
-def _write_version_file(project_name, latest_version):
-    file_name = '{0}.version'.format(project_name)
-    full_path = builder_util.RUNTIME_BUCKET_PREFIX + file_name
-    builder_util.write_to_gcs(full_path, latest_version)
+        parts = os.path.splitext(latest)
+        project_prefix = builder_util.RUNTIME_BUCKET_PREFIX \
```
Comment (dlorenc) [inline runtime_builders/publish_builders.py]: Could we use path.join instead of + here, to be safe with /'s

File: runtime_builders/builder_util.py (hunk 17 above)
Comment (nkubala) [inline runtime_builders/builder_util.py]: done

File: runtime_builders/publish_builders.py (hunk 18 above)
Comment (nkubala) [inline runtime_builders/publish_builders.py]: ah yeah. that's clean, but in the end I'd just be moving this same logic to another helper method, so the effect is basically the same. good to know for the future though

File: runtime_builders/publish_builders.py (hunk 19 above)
Comment (nkubala) [inline runtime_builders/publish_builders.py]: done
